File: frontend/src/lib/services/bead.service.ts
/**
 * Bead Service - API client for beads endpoints
 */
import { dedupedFetch } from '../api/client';
import type { Bead } from '../types/shared-types';

export interface BeadCreate {
  parent_id?: string;
  branch_name: string;
  action: string;
  emotion_tag?: string;
  content: Record<string, any>;
  signature?: string;
}

export interface BeadResponse extends Bead {
  timestamp: string;
}

export interface BranchCreate {
  branch_name: string;
  from_bead_id: string;
}

export interface BranchResponse {
  branch_name: string;
  head_bead_id: string;
  message: string;
}

export interface BeadDiffChange {
  field: string;
  from: any;
  to: any;
}

export interface BeadDiffResponse {
  bead_id_earlier: string;
  bead_id_later: string;
  changes: BeadDiffChange[];
}

class BeadService {
  private basePath = '/api/v1/beads';

  async getTimeline(
    branch?: string,
    limit: number = 100,
    offset: number = 0
  ): Promise<BeadResponse[]> {
    const params = new URLSearchParams({
      limit: limit.toString(),
      offset: offset.toString(),
      ...(branch && { branch })
    });
    return dedupedFetch<BeadResponse[]>(`${this.basePath}/timeline?${params}`, {
      cacheKey: `beads:timeline:${branch || 'main'}:${limit}:${offset}`,
      cacheTtl: 60000 // 1 minute
    });
  }

  async create(data: BeadCreate): Promise<BeadResponse> {
    return dedupedFetch<BeadResponse>(this.basePath, {
      method: 'POST',
      body: JSON.stringify(data)
    });
  }

  async createBranch(branchName: string, fromBeadId: string): Promise<BranchResponse> {
    return dedupedFetch<BranchResponse>(`${this.basePath}/branch`, {
      method: 'POST',
      body: JSON.stringify({ branch_name: branchName, from_bead_id: fromBeadId })
    });
  }

  async diff(beadId1: string, beadId2: string): Promise<BeadDiffResponse> {
    return dedupedFetch<BeadDiffResponse>(`${this.basePath}/diff/${beadId1}/${beadId2}`);
  }

  async listBranches(): Promise<{ branches: Array<{ name: string; head_bead_id: string; head_timestamp: string }> }> {
    return dedupedFetch(`/${this.basePath}/branches`, {
      cacheKey: 'branches:all',
      cacheTtl: 120000 // 2 minutes
    });
  }

  async getBead(id: string): Promise<BeadResponse> {
    // Individual bead fetch not in API yet
    throw new Error('Not implemented');
  }
}

export const beadService = new BeadService();

File: backend/services/bead_service.py
"""
Bead Service for managing Beads (DAG nodes).
"""

from typing import Dict, List, Optional, Any
import hashlib
from datetime import datetime


class BeadService:
    """Service for managing beads in the narrative DAG."""

    def __init__(self):
        """Initialize BeadService."""
        self._beads: Dict[str, Dict[str, Any]] = {}

    async def create_bead(
        self,
        parent_id: Optional[str],
        content: Dict[str, Any],
        action: str,
        branch_name: str = "main",
        emotion_tag: Optional[str] = None,
    ) -> str:
        """Create a new bead in the DAG."""
        bead_data = {
            "parent_id": parent_id,
            "content": content,
            "action": action,
            "branch_name": branch_name,
            "emotion_tag": emotion_tag,
            "timestamp": datetime.utcnow().isoformat(),
        }
        bead_id = hashlib.sha1(str(bead_data).encode()).hexdigest()[:40]
        bead_data["id"] = bead_id
        self._beads[bead_id] = bead_data
        return bead_id

    async def get_bead(self, bead_id: str) -> Optional[Dict[str, Any]]:
        """Get a bead by ID."""
        return self._beads.get(bead_id)

    async def get_timeline(
        self, branch_name: str = "main", limit: int = 50
    ) -> List[Dict[str, Any]]:
        """Get timeline for a branch."""
        beads = [b for b in self._beads.values() if b.get("branch_name") == branch_name]
        beads.sort(key=lambda x: x.get("timestamp", ""))
        return beads[:limit]

    async def create_branch(self, branch_name: str, from_bead_id: str) -> str:
        """Create a new branch from an existing bead."""
        return f"branch_{branch_name}_{from_bead_id[:8]}"

    async def merge_branches(self, source_branch: str, target_branch: str) -> str:
        """Merge two branches."""
        return hashlib.sha1(
            f"merge_{source_branch}_{target_branch}".encode()
        ).hexdigest()[:40]
